rlog/2024-04-20-rln-light-verifiers.mdx

```diff
@@ -0,0 +1,173 @@
+- Hash function: PoseidonT3 (which is a gas guzzler)
+- Max size of the membership set: 2^20 = 1,048,576 leaves
+
+### Metrics
+
+- Time to sync the whole tree: 4 minutes
+- RPC calls: 702
+- Number of leaves: 2653
+
+One can argue that the time to sync the tree at the current state is not _that_ bad.
+However, the number of RPC calls is a concern,
+which scales linearly with the number of blocks since the contract was deployed.
+This is because the implementation fetches all events from the contract,
+chunking 2000 blocks at a time.
+This is done to avoid hitting the block limit of 10,000 events per call,
+which is a limitation of popular RPC providers.
+
+## Proposed solution
+
+From a theoretical perspective,
+one could construct the merkle tree on-chain,
+in a view call, in-memory.
+However, this is not feasible due to the gas costs associated with it.
+
+To compute the root of a Merkle tree with $2^{20}$ leaves,
+it costs approximately 2 Billion Gas.
+With Infura and Alchemy capping the gas limit to 350M and 550M gas respectively,
+it is not possible to compute the root of the tree in a single call.
+
+Acknowledging that [Polygon Miden](https://polygon.technology/blog/polygon-miden-state-model) and [Penumbra](https://penumbra.zone/blog/tiered-commitment-tree/) both make use of a tiered commitment tree,
+we propose a similar approach for RLN.
+
+A tiered commitment tree is a tree which is sharded into multiple smaller subtrees,
+each of which is a tree in itself.
+This allows scaling in terms of the number of leaves,
+as well as reducing state bloat,
+by just storing the root of a subtree when it is full,
+instead of all its leaves.
+
+Here, the main question arises:
+What is the maximum number of leaves in a subtree, with which the root can be computed in a single call?
+
+To compute the root of a Merkle tree with $2^{10}$ leaves, it costs approximately 217M gas.
+
+This is a feasible number for a single call,
+and hence we propose a tiered commitment tree with a maximum of $2^{10}$ leaves in a subtree,
+and the number of subtrees is $2^{10}$.
+Therefore, the maximum number of leaves in the tree is $2^{20}$,
+the same as the current implementation.
+
+![img](/img/light-rln-verifiers.png)
+
+### Insertion
+
+When a commitment is inserted into the tree,
+it is first inserted into the first subtree.
+When the first subtree is full,
+the next insertions go into the second subtree,
+and so on.
+
+### Syncing
+
+When syncing the tree,
+one only needs to fetch the roots of the subtrees,
+and the root of the tree can be computed in-memory, or on-chain.
+
+Which allows us to derive the following relation:
+
+$$
+number\_of\_rpc\_calls = number\_of\_filled\_subtrees + 1
+$$
+
+This is a significant improvement over the current implementation,
+which requires fetching all the memberships from the smart contract.
+
+### Gas costs
+
+The gas costs for inserting a commitment into the tree are the same as the current implementation,
```

Comment:
Drop comma between 'implementation' and 'except'

Reply:
addressed in 567b53f